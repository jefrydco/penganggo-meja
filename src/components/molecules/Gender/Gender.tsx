import { GENDER_LIST } from "@/constants/gender";
import { GenderEnum } from "@/interfaces/user";
import {
  Box,
  FormControl,
  InputLabel,
  MenuItem,
  Select,
  SelectChangeEvent,
} from "@mui/material";
import React, { useEffect, useState } from "react";

export interface GenderProps {
  reset?: boolean
  onGenderChange?: ((gender: GenderEnum) => void) | undefined;
}

export const Gender: React.FC<GenderProps> = ({ reset, onGenderChange }) => {
  const [gender, setGender] = useState<GenderEnum>(GenderEnum.FEMALE);
  const handleGenderChange:
    | ((event: SelectChangeEvent<GenderEnum>, child: React.ReactNode) => void)
    | undefined = (event) => {
    const newGender = event.target.value as GenderEnum;
    setGender(newGender);
    onGenderChange && onGenderChange(newGender);
  };
  useEffect(() => {
    if (reset) {
      setGender(GenderEnum.FEMALE)
    }
  }, [reset])
  return (
    <FormControl variant="filled">
      <InputLabel id="gender-label">Gender</InputLabel>
      <Select
        data-testid="gender"
        labelId="gender-label"
        id="gender-select"
        value={gender}
        onChange={handleGenderChange}
      >
        {GENDER_LIST.map((genderItem) => (
          <MenuItem key={genderItem} value={genderItem}>
            {genderItem}
          </MenuItem>
        ))}
      </Select>
    </FormControl>
  );
};
