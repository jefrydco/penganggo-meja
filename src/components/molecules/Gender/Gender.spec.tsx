import React from "react";
import { describe, expect, it, vi } from "vitest";
import {
  getByRole,
  render,
  screen,
  userEvent,
  waitFor,
} from "@/utils/tests/vitest";
import { Gender } from "./Gender";

describe("Gender", () => {
  it("match snapshot", () => {
    const { asFragment } = render(<Gender />);
    expect(asFragment()).toMatchSnapshot();
  });
  it("clickable", () => {
    const handleGenderChange = vi.fn();
    const user = userEvent.setup();
    const { getByTestId, getByRole } = render(
      <Gender onGenderChange={handleGenderChange} />
    );
    user.click(getByTestId("gender"));
    user.tab();
    user.type(getByTestId("gender"), "{arrowdown}");
    user.type(getByTestId("gender"), "{enter}");
    waitFor(() => {
      expect(getByRole("button")).toHaveTextContent('male');
      expect(handleGenderChange).toBeCalledTimes(1);
    });
  });
});
