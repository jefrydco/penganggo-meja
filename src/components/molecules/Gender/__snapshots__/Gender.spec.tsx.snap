// Vitest Snapshot v1

exports[`Gender > match snapshot 1`] = `
<DocumentFragment>
  <div
    class="MuiFormControl-root css-1nrlq1o-MuiFormControl-root"
  >
    <label
      class="MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated MuiInputLabel-shrink MuiInputLabel-filled MuiFormLabel-root MuiFormLabel-colorPrimary MuiFormLabel-filled css-o943dk-MuiFormLabel-root-MuiInputLabel-root"
      data-shrink="true"
      id="gender-label"
    >
      Gender
    </label>
    <div
      class="MuiFilledInput-root MuiFilledInput-underline MuiInputBase-root MuiInputBase-colorPrimary MuiInputBase-formControl  css-67qocj-MuiInputBase-root-MuiFilledInput-root-MuiSelect-root"
      data-testid="gender"
    >
      <div
        aria-expanded="false"
        aria-haspopup="listbox"
        aria-labelledby="gender-label gender-select"
        class="MuiSelect-select MuiSelect-filled MuiFilledInput-input MuiInputBase-input css-d9oaum-MuiSelect-select-MuiInputBase-input-MuiFilledInput-input"
        id="gender-select"
        role="button"
        tabindex="0"
      >
        female
      </div>
      <input
        aria-hidden="true"
        class="MuiSelect-nativeInput css-yf8vq0-MuiSelect-nativeInput"
        tabindex="-1"
        value="female"
      />
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium MuiSelect-icon MuiSelect-iconFilled css-hfutr2-MuiSvgIcon-root-MuiSelect-icon"
        data-testid="ArrowDropDownIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M7 10l5 5 5-5z"
        />
      </svg>
    </div>
  </div>
</DocumentFragment>
`;
